File: python/dsa/data-structures/stack-queue/remove-k-digits.py
def removeKdigits(numbers: str, kTotal: int) -> str:
    resStack = [] # @viz: init_stack
    
    for digit in numbers: # @viz: loop
        while kTotal > 0 and resStack and resStack[-1] > digit: # @viz: while_check
            resStack.pop() # @viz: pop
            kTotal -= 1 # @viz: dec_k
        resStack.append(digit) # @viz: push
        
    while kTotal > 0 and resStack: # @viz: trim_check
        resStack.pop() # @viz: pop_trim
        kTotal -= 1 # @viz: dec_k_trim

    resStr = "".join(resStack).lstrip('0') # @viz: join
    return resStr if resStr else "0" # @viz: finish

File: cpp/dsa/data-structures/stack-queue/remove-k-digits.cpp
class Solution {
public:
    string removeKdigits(string numString, int kValue) {
        string resContainer = ""; // @viz: init_stack
        
        for (char digit : numString) { // @viz: loop
            while (kValue > 0 && !resContainer.empty() && resContainer.back() > digit) { // @viz: while_check
                resContainer.pop_back(); // @viz: pop
                kValue--; // @viz: dec_k
            }
            resContainer.push_back(digit); // @viz: push
        }
        
        while (kValue > 0) { // @viz: trim_check
            resContainer.pop_back(); // @viz: pop_trim
            kValue--; // @viz: dec_k_trim
        }
        
        size_t firstIdx = resContainer.find_first_not_of('0'); // @viz: join
        if (firstIdx == string::npos) return "0"; // @viz: finish
        return resContainer.substr(firstIdx);
    }
}